mllib/src/main/scala/org/apache/spark/ml/image/ImageSchema.scala

```diff
@@ -0,0 +1,229 @@
+  private val imageDFSchema = StructType(StructField("image", columnSchema, true) :: Nil)
+
+  @Since("2.3.0")
+  def getOrigin(row: Row): String = row.getString(0)
+
+  @Since("2.3.0")
+  def getHeight(row: Row): Int = row.getInt(1)
+
+  @Since("2.3.0")
+  def getWidth(row: Row): Int = row.getInt(2)
+
+  @Since("2.3.0")
+  def getNChannels(row: Row): Int = row.getInt(3)
+
+  @Since("2.3.0")
+  def getMode(row: Row): String = row.getString(4)
+
+  @Since("2.3.0")
+  def getData(row: Row): Array[Byte] = row.getAs[Array[Byte]](5)
+
+  /**
+   * Check if the dataframe column contains images (i.e. has ImageSchema)
+   *
+   * @param df       Dataframe
+   * @param column   Column name
+   * @return         True if the given column matches the image schema
+   */
+  @Since("2.3.0")
+  def isImageColumn(df: DataFrame, column: String): Boolean =
+    df.schema(column).dataType == columnSchema
+
+  /**
+   * Default values for the invalid image
+   *
+   * @param origin Origin of the invalid image
+   * @return       Row with the default values
+   */
+  private def invalidImageRow(origin: String): Row = Row(Row(origin, -1, -1, -1, undefinedImageType,
+    Array.ofDim[Byte](0)))
+
+  /**
+   * Convert the compressed image (jpeg, png, etc.) into OpenCV
+   * representation and store it in dataframe Row
+   *
+   * @param origin   Arbitrary string that identifies the image
+   * @param bytes    Image bytes (for example, jpeg)
+   * @return         Dataframe Row or None (if the decompression fails)
+   */
+  private[spark] def decode(origin: String, bytes: Array[Byte]): Option[Row] = {
+
+    val img = ImageIO.read(new ByteArrayInputStream(bytes))
+
+    if (img == null) {
+      None
+    } else {
+      val is_gray = img.getColorModel.getColorSpace.getType == ColorSpace.TYPE_GRAY
+      val has_alpha = img.getColorModel.hasAlpha
+
+      val height = img.getHeight
+      val width = img.getWidth
+      val (nChannels, mode) = if (is_gray) {
+        (1, "CV_8UC1")
+      } else if (has_alpha) {
+        (4, "CV_8UC4")
+      } else {
+        (3, "CV_8UC3")
+      }
+
+      assert(height * width * nChannels < 1e9, "image is too large")
+      val decoded = Array.ofDim[Byte](height*width*nChannels)
+
+      // Grayscale images in Java require special handling to get the correct intensity
+      if (is_gray) {
+        var offset = 0
+        val raster = img.getRaster
+        for(h <- 0 until height) {
+          for (w <- 0 until width) {
+            decoded(offset) = raster.getSample(w, h, 0).toByte
+            offset += 1
+          }
+        }
+      }
+      else {
+        var offset = 0
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            val color = new Color(img.getRGB(w, h))
+
+            decoded(offset) = color.getBlue.toByte
+            decoded(offset + 1) = color.getGreen.toByte
+            decoded(offset + 2) = color.getRed.toByte
```

Comment:
Will overflow be an issue later?

scala> val c = new Color(255, 255, 255)
c: java.awt.Color = java.awt.Color[r=255,g=255,b=255]

scala> c.getGreen
res12: Int = 255

scala> c.getGreen.toByte
res13: Byte = -1

Reply:
Yeah any code reading the byte must treat them as unsigned, but this isn't necessarily an error

Reply:
I agree. The result is signed, but it is not an error for the Byte type.